edge/pkg/edged/volume/csi/csi_attacher.go

```diff
@@ -0,0 +1,447 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	storage "k8s.io/api/storage/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const (
+	persistentVolumeInGlobalPath = "pv"
+	globalMountInGlobalPath      = "globalmount"
+)
+
+type csiAttacher struct {
+	plugin        *csiPlugin
+	k8s           kubernetes.Interface
+	waitSleepTime time.Duration
+
+	csiClient csiClient
+}
+
+// volume.Attacher methods
+var _ volume.Attacher = &csiAttacher{}
+
+var _ volume.Detacher = &csiAttacher{}
+
+var _ volume.DeviceMounter = &csiAttacher{}
+
+func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
+	return "", nil
+}
+
+func (c *csiAttacher) WaitForAttach(spec *volume.Spec, _ string, pod *v1.Pod, timeout time.Duration) (string, error) {
+	source, err := getPVSourceFromSpec(spec)
+	if err != nil {
+		klog.Error(log("attacher.WaitForAttach failed to extract CSI volume source: %v", err))
+		return "", err
+	}
+
+	attachID := getAttachmentName(source.VolumeHandle, source.Driver, string(c.plugin.host.GetNodeName()))
+
+	return c.waitForVolumeAttachment(source.VolumeHandle, attachID, timeout)
+}
+
+func (c *csiAttacher) waitForVolumeAttachment(volumeHandle, attachID string, timeout time.Duration) (string, error) {
+	klog.V(4).Info(log("probing for updates from CSI driver for [attachment.ID=%v]", attachID))
+
+	err := wait.PollImmediate(time.Second*5, time.Minute*5, func() (bool, error) {
+		klog.V(4).Info(log("probing VolumeAttachment [id=%v]", attachID))
+		attach, err := c.k8s.StorageV1().VolumeAttachments().Get(attachID, meta.GetOptions{})
+		if err != nil {
+			klog.Error(log("attacher.WaitForAttach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
+			return false, fmt.Errorf("volume %v has GET error for volume attachment %v: %v", volumeHandle, attachID, err)
```

Comment:
I think we should either use, logging directly OR return an error. In this case I would prefer the second option.

Reply:
Thanks. I will

Reply:
+1

Reply:
Done. Thank you